### DISCUSSION TITLE
createStore on nextSSR with nx lib

### DISCUSSION THREAD

**Discussion by mmswi** [pmndrs/jotai#2032] [Q&A] [answered]
Hello guys,

Congrats and thank you for the work! I'm having a question for a better understanding of Jotai, but let me explain the scenario.

I'm having a next > 13 with nx, project. The stores are created in a shared lib, `libs/shared/stores/src/lib/[StoreA.ts, StoreB.ts, etc]`

In each of these files, I have an export
`export const myStoreA = createStore()`  with `myStoreA.set(myAtom, null)`

Then in my components, I'm directly importing this `myStoreA` from that shared lib and use it to `set/get` values

So, in MyComponentA.tsx -> `import {MyStoreA} from '@myprj/shared/stores'` then in the component `MyStoreA.set(myAtom, 'some-new-value')`. The same in other components, I might import the store and read that value, or `useAtomValue(myAtom, {store: MyStoreA})`

In `app.tsx`, over the whole app, I've wrapped the Jotai `<Provider><MyAPP /></Provider>`

The application is **SSR**.

My question is, do I have any risks of **memory leaks** using the above approach? 
Should I create a new instance of the stores and pass it in a custom provider, when the MyApp loads, or I can just create it in an nx lib and import it from there?

ex for creating the instance:

      const myStoreA = createStoreAInstance() //returning a myStoreA = createStore() with myStoreA.set(myAtom, null)

      export const myApp = () => {
          return <MyCustomStoreAProvider store={myStoreA}> <MyApp> </MyCustomStoreAProvider>

and then using a hook in all my components, `const myStoreA = useStoreA()`?

Appreciate the response

**Comment by dai-shi** [comment]
> My question is, do I have any risks of **memory leaks** using the above approach?

If you use `<Provider><MyAPP /></Provider>`, the store will be garbage collected after unmounting.

Check out the source code. It's tiny:
https://github.com/pmndrs/jotai/blob/4f8d1feb7ae67097b37baa3fd72036776f6067f2/src/react/Provider.ts#L18-L36

**Reply by mmswi** [reply]
Thanks for the answer. My only confusion is this.
I have 3 stores and I wrap the app in `<Provider><MyAPP /></Provider>`, to me it seems that it creates the `defaultStore` but it still keeps the other 3 stores uncollected.  
Do I get this correctly?

I'm passing in my components, `import {myStoreA}`, `import {myStoreB}`,` import {MyStoreC}`. Then I'm doing
`useAtomValue(myAtomA, store: myStoreA}`

`myStoreB.set()`

`myStoreC.get()`

etc for many stores.

Will these get garbage collected? Because to me it seems they won't. And the risk that I see is that the nx lib might cache or keep some references to those exported store constants that I import in my components. 

Disclaimer, I'm not a backend engineer and don't have that much experience with nx, just have some concerns.

Thanks for explaining!

**Reply by dai-shi** [reply, marked as answer]
I misunderstood. You use `<Provider store={...}>` instead of just `<Provider>`. In such a case, it's not garbage collected unless you clear your store references.

**Reply by mmswi** [reply]
Thank you for the answers! This was my thought exactly, but needed a second opinion.
